### PR TITLE
chore: adds consistent vanity import to files and provides tooling for verifying and updating them.

### LINKED ISSUE #120641: Automate vanity imports update and verify
### What would you like to be added?

Vanity imports to all packages, for example

```
package validatingwebhookconfiguration
```
into
```
package validatingwebhookconfiguration // import "k8s.io/kubernetes/pkg/registry/admissionregistration/validatingwebhookconfiguration"
```

**IMPORTANT:** Kubernetes repository already uses vanity imports, we just need to provide a check so new code includes it and provide a tool for users to regenerate vanity imports.

### Why is this needed?

Following https://github.com/kubernetes/kubernetes/issues/68231, vanity imports improves code navegability for a codebase with many packages as otherwise you need to figure out what the package is, the vanity URL and probably handtype the import path.

### PR DESCRIPTION
**Pull request by jcchavezs**
#### What type of PR is this?
/kind documentation

#### What this PR does / why we need it:

This PR adds vanity import to all the exported code in this repository by using the tool porto. Vanity imports are very helpful when it comes to working with

- libraries with many packages (and long import names)
- libraries with non CVS's domain (e.g. k8s.io instead of github.com/kubernetes) as it tells the user what is the import to use for a package without the need of copying the fullpath and guessing the domain of the module. For example, imagine checking the code of this file and having to resolve the import:
```
package validatingwebhookconfiguration
```
into
```
package validatingwebhookconfiguration // import "k8s.io/kubernetes/pkg/registry/admissionregistration/validatingwebhookconfiguration"
```

#### Which issue(s) this PR fixes:

Fixes #120641

#### Special notes for your reviewer:

**IMPORTANT:** Kubernetes repository already uses vanity imports, we just provide a check so new code includes it and provide a tool for users to regenerate vanity imports.

Also **the important changes belong to the `hack` folder** (see https://github.com/kubernetes/kubernetes/pull/120642/files?file-filters%5B%5D=.sh&show-viewed-files=true), all the rest are generated.

As for the chosen tool, OpenTelemetry already uses this (see https://github.com/open-telemetry/opentelemetry-go/blob/main/Makefile#L69)

#### Does this PR introduce a user-facing change?
NONE

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: hack/make-rules/update.sh [hunk 1]
```diff
@@ -40,6 +40,7 @@ BASH_TARGETS=(
 	update-generated-api-compatibility-data
 	update-generated-docs
 	update-openapi-spec
+	update-vanity-imports
```
Comment (jcchavezs) [inline hack/make-rules/update.sh:43]: Important change: adds vanity imports to the `update.sh`

File: hack/make-rules/verify.sh [hunk 2]
```diff
@@ -74,6 +74,7 @@ QUICK_PATTERNS+=(
   "verify-gofmt.sh"
   "verify-imports.sh"
   "verify-non-mutating-validation.sh"
+  "verify-vanity-imports.sh"
```
Comment (jcchavezs) [inline hack/make-rules/verify.sh:77]: Important change: adds vanity imports to the `verify.sh`

File: hack/update-vanity-imports.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
```
Comment (jcchavezs) [inline hack/update-vanity-imports.sh:1]: This is the script that updates all vanity imports.

File: hack/verify-vanity-imports.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
```
Comment (jcchavezs) [inline hack/verify-vanity-imports.sh:1]: This is the script that verifies all vanity imports.

**Comment by liggitt** [comment]
I agree with @neolit123's comment at https://github.com/kubernetes/kubernetes/issues/120641#issuecomment-1721729413 that this is too noisy and not necessary. Adding one more thing presubmits can fail on and another formatting script contributors need to run needs to be solving a clear problem to justify the toil/churn, and I don't see that here.

**Comment by jcchavezs** [comment]
@liggitt @neolit123, I am happy to change this PR so that only doc.go files are affected, are you OK with that?

File: hack/make-rules/verify.sh (hunk 2 above)
Comment (liggitt) [reply hack/make-rules/verify.sh:77]: this is the main part I disagree with... I don't think vanity import comments are so important we should fail or block PRs on them

**Comment by liggitt** [comment]
> @liggitt @neolit123, I am happy to change this PR so that only doc.go files are affected, are you OK with that?

scoping to just doc.go files and dropping the verify script so we don't block on this would make it more acceptable to me

**Comment by jcchavezs** [comment]
Waiting for https://github.com/jcchavezs/porto/pull/17

**Comment by jcchavezs** [comment]
@liggitt I just updated the PR scoping the update to `doc.go` only.

**Comment by jcchavezs** [comment]
Friendly ping @liggitt

File: hack/make-rules/verify.sh (hunk 2 above)
Comment (liggitt) [reply hack/make-rules/verify.sh:77]: I'm still -1 on making vanity imports a hard requirement that will fail pull requests. I don't think they are necessary or important enough to increase contributor friction over.

Comment (liggitt) [reply hack/make-rules/verify.sh:77]: I also think the autogeneration should be scoped to things under staging/src/k8s.io/...

Things under k8s.io/kubernetes/... outside of staging are not intended for use as a library.

**Comment by jcchavezs** [comment]
Friendly ping @liggitt

File: cmd/kubeadm/app/cmd/options/doc.go [hunk 5]
```diff
@@ -26,4 +26,4 @@ In addition to defining the flags, the package also contains set of utilities fo
 For additional details about how flags are managed in phases, please refer to the
 "k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow" package.
 */
-package options
+package options // import "k8s.io/kubernetes/cmd/kubeadm/app/cmd/options"
```
Comment (liggitt) [inline cmd/kubeadm/app/cmd/options/doc.go:29]: revert the changes outside staging (cmd/... and pkg/...)

File: hack/make-rules/verify.sh (hunk 2 above)
Comment (liggitt) [reply hack/make-rules/verify.sh:77]: > If you check https://github.com/kubernetes/kubernetes/pull/120642/files#diff-e0ed342fde1ba5da0d2c7dcb506308e906b35c6b527c5150ec157da9b89b5e69R34-R35 we are just inspecting and showing a message, not making the PR fail. I can remove it too if you prefer so.

Ah, I missed that. Yeah, I don't think many people run this and inspect the results unless something in CI fails, let's drop this bit.

File: hack/update-vanity-imports.sh [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "${KUBE_ROOT}/hack/lib/init.sh"
+
+kube::golang::verify_go_version
+
+GO111MODULE=on go install github.com/jcchavezs/porto/cmd/porto@v0.5.1
```
Comment (liggitt) [inline hack/update-vanity-imports.sh:29]: as much as possible, we try to keep tool dependencies recorded in `hack/tools/go.mod` and install them like

```
pushd "${KUBE_ROOT}/hack/tools" >/dev/null
  GO111MODULE=on go install github.com/jcchavezs/porto/cmd/porto
popd >/dev/null
```

Comment (jcchavezs) [reply hack/update-vanity-imports.sh:29]: Done.

**Comment by jcchavezs** [comment]
PTAL @liggitt

**Comment by jcchavezs** [comment]
@liggitt could we attempt to merge this before I have to rebase it again please?

File: hack/make-rules/update.sh (hunk 1 above)
Comment (liggitt) [reply hack/make-rules/update.sh:43]: drop the call here, everything else in this list is updating things enforced by verify scripts, but including this will force changes to possibly irrelevant files when people are making changes like API updates

Comment (jcchavezs) [inline hack/make-rules/update.sh:43]:
```suggestion
```

Comment (jcchavezs) [reply hack/make-rules/update.sh]: Done.

**Comment by jcchavezs** [comment]
@liggitt the failure is unrelated https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120642/pull-kubernetes-verify/1725949438721003520 what do you think?

**Comment by liggitt** [comment]
the failure is due to this PR:

```
In ./hack/update-vanity-imports.sh line 33:
porto --restrict-to-dirs="staging" --restrict-to-files="doc\\.go$" -w ${KUBE_ROOT}
                                                                      ^----------^ SC2086 (info): Double quote to prevent globbing and word splitting.
```

**Comment by pacoxu** [comment]
/release-note-none
/lgtm cancel
as the verify CI needs to be fixed.

**Comment by bart0sh** [comment]
@jcchavezs please address comments and fix failing CI jobs, thanks.

**Comment by jcchavezs** [comment]
I will do that.

**Comment by jcchavezs** [comment]
@bart0sh I just did the rebase.

**Comment by jcchavezs** [comment]
Can we retrigger the failing test?

**Comment by jcchavezs** [comment]
This PR is green @leilajal
